File: 0003-longest-substring-without-repeating-characters/solution.rs
impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut last_seen: [Option<usize>; 128] = [None; 128];
        let mut max_length: usize = 0;
        let mut start: usize = 0;
        for (index, char) in s.char_indices() {
            let byte = char as usize;
            if let Some(prev_index) = last_seen[byte] {
                if prev_index >= start {
                    start = prev_index + 1;
                }
            }
            last_seen[byte] = Some(index);
            max_length = max_length.max(index - start + 1);
        }
        max_length as i32
    }
}

File: 0003-longest-substring-without-repeating-characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        last_seen = [None] * 128  # Assuming ASCII
        result = 0
        start = 0

        for index, char in enumerate(s):
            prev_index = last_seen[ord(char)]
            if prev_index is not None and prev_index >= start:
                start = prev_index + 1
            last_seen[ord(char)] = index
            result = max(result, index - start + 1)

        return result


